py/server/deephaven/experimental/pivot.py

```diff
@@ -0,0 +1,108 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+
+"""This module defines functions for creating pivot tables."""
+
+from typing import Sequence, Union, Optional, Callable, Any
+import re
+from deephaven import DHError, empty_table
+from deephaven.table import Table, PartitionedTable, multi_join
+from deephaven.numpy import to_numpy
+from deephaven.update_graph import auto_locking_ctx
+from deephaven.jcompat import to_sequence
+
+
+def _is_legal_column(s: str) -> bool:
+    """Check if a column name is legal.
+
+    Args:
+        s (str): The column name to check.
+
+    Returns:
+        bool: True if the column name is legal, False otherwise.
+    """
+    return re.match("^[_a-zA-Z][_a-zA-Z0-9]*$", s) is not None
+
+
+def _legalize_column(s: str) -> str:
+    """Legalize a column name.  Invalid characters are replaced with underscores.
+    The legalized column name is not guaranteed to be unique.
+
+    Args:
+        s (str): The column name to legalize.
+
+    Returns:
+        str: The legalized column name.
+
+    Raises:
+        ValueError: If the column name is empty.
+    """
+    if re.match("^[_a-zA-Z][_a-zA-Z0-9]*$", s):
+        return s
+    if re.match("^[_a-zA-Z].*$", s):
+        return re.sub("[^_a-zA-Z0-9]", "_", s)
+    return "_" + re.sub("[^_a-zA-Z0-9]", "_", s)
+
+
+def pivot(table: Table, row_cols: Union[str, Sequence[str]], column_col: str, value_col: str,
+          value_to_col_name: Optional[Callable[[Any], str]] = None) -> Table:
```

Comment:
I would assume for true pivot tables, customers will ask to be able to pass in an array for columns and values as well, similar to Excel: https://support.microsoft.com/en-us/office/pivotby-function-de86516a-90ad-4ced-8522-3a25fac389cf